### PR TITLE
Use enums instead of booleans in TransportReplicationAction

### PR DESCRIPTION
**Pull request by nicktindall**
In preparation for #109414, so when we add a third behavioural flag it doesn't get unwieldy.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 1]
```diff
@@ -89,6 +89,34 @@ public abstract class TransportReplicationAction<
...
+    protected enum SyncGlobalCheckpointAfterOperation {
+        /**
+         * Do not sync as part of this action
+         */
+        DoNotSync,
+        /**
+         * Attempt to sync the global checkpoint to the replica(s) after success
+         */
+        AttemptAfterSuccess
+    }
```
Comment (nicktindall) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:118]: Names are obviously up for discussion. Assumed these will probably always be a binary choice between two options, but wanted to make them somewhat descriptive and leave the door open for more to be added if need be.

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 2]
```diff
@@ -89,6 +89,34 @@ public abstract class TransportReplicationAction<
...
     Response extends ReplicationResponse> extends TransportAction<Request, Response> {
 
+    /**
+     * Execution of the primary action
+     */
+    protected enum PrimaryActionExecution {
+        /**
+         * Is subject to usual queue length and indexing pressure checks
+         */
+        Normal,
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:99]:
```suggestion
        RejectOnOverload,
```

File: server/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java [hunk 3]
```diff
@@ -1511,8 +1511,8 @@ private class TestAction extends TransportReplicationAction<Request, Request, Te
                 Request::new,
                 Request::new,
                 EsExecutors.DIRECT_EXECUTOR_SERVICE,
-                false,
-                forceExecute
+                SyncGlobalCheckpointAfterOperation.DoNotSync,
+                forceExecute ? PrimaryActionExecution.Force : PrimaryActionExecution.Normal
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/action/support/replication/TransportReplicationActionTests.java:1515]: Could we propagate this back through the tests as well?

File: server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java [hunk 4]
```diff
@@ -187,7 +187,10 @@ protected TransportReplicationAction(
 
         this.initialRetryBackoffBound = REPLICATION_INITIAL_RETRY_BACKOFF_BOUND.get(settings);
         this.retryTimeout = REPLICATION_RETRY_TIMEOUT.get(settings);
-        this.forceExecutionOnPrimary = forceExecutionOnPrimary;
+        this.forceExecutionOnPrimary = switch (primaryActionExecution) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java:190]: Suggests we should propagate this change down into the transport layer too (of which I'd be supportive). Maybe in a followup.

**Review by DaveCTurner — COMMENTED** [review]
Looks good, a couple of comments and a naming suggestion.

**Review by idegtiarenko — APPROVED** [review]

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**